proposals/p1885.md

```diff
@@ -657,6 +701,91 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
 These two options could allow interoperability with a compatible C++ type,
 either implicitly (template approach) or explicitly (adapter approach).
 
+#### Iterating over Carbon types in C++
+
+C++ needs to provide `begin()` and `end()` methods or free functions, in
+addition to `*` and `++` operators.
+
+The `Iterate` interface does not expose a way to have a cursor nor an iterator
+to the past-last element. An option would be to have `end()` return a predefined
+invalid value, to satisfy iterator comparison when `Next()` returns an empty
+`Optional`.
+
+Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
+need to copy and cache the last value, to allow for `*it` to be called. This
+puts a requirement on `ElementType` to be copyable for this interoperability to
+be usable.
+
+```cpp
+namespace Carbon {
+// A C++ input iterator for any type that implements the Carbon `Iterate` interface.
+template <typename Iterate>
+class IterateIterator {
+ public:
+  // Returns an "Invalid" iterator representing `end()`
+  static auto Invalid(const Iterate& container) -> IterateIterator<Iterate> {
+    return IterateIterator<Iterate>(container, std::nullopt);
+  }
+
+  IterateIterator(const Iterate& container,
+                  std::optional<typename Iterate::CursorType> cursor)
+      : container_(&container), cursor_(cursor) {
+    if (cursor_) {
+      next();
+    }
+  }
+  IterateIterator(const IterateIterator& other)
+      : container_(other.container_), cursor_(other.cursor_),
+      value_(other) {}
+
+  auto operator*() const -> const typename Iterate::ElementType& {
+    assert(cursor_);
+    return value_;
+  }
+  auto operator++() -> IterateIterator<Iterate>& {
```

Comment:
Probably should have both this and the version that takes an `int` parameter, so this iterator supports both pre- and post- increment.

Reply:
I don't think I agree. Post-increment won't work if `ElementType` isn't copyable. By the same token, this type won't be a valid iterator if `ElementType` isn't copyable, but it will still work with range-based for loops, so I think we should focus solely on the features needed to support that.

Reply:
Agree with @geoffromer , resolving this one as I don't think this would be a blocking issue. Feel free to re-open if you think otherwise @josh11b .